stable_baselines/gail/dataset/record_expert.py

```diff
@@ -101,22 +104,32 @@ def generate_expert_traj(model, save_path, env=None, n_timesteps=0,
             observations.append(obs)
 
         if isinstance(model, BaseRLModel):
-            action, _ = model.predict(obs)
+            action, state = model.predict(obs, state=state)
```

Comment:
After looking at the code, the right way to do that is:
```python
done = False
if is_vec_env:
    done = [False for _ in range(len(env.num_envs)]
...
   action, state = model.predict(obs, state=state, mask=done)
```
and you can remove state = None in the `if done`

Reply:
@hill-a @erniejunior @AdamGleave weirdly enough, during my tests, when mask is not passed, I get better results...

Reply:
https://github.com/hill-a/stable-baselines/blob/e3908df6615ed20c7be4f7968a392865a68ab4c6/stable_baselines/common/base_class.py#L463-L467

I disagree you need to keep track of the `state`, but `mask` is automatically set to `False`, if `mask` is set to `None` (what it is by default). I think by not creating a `mask` in `record_expert` the code is a bit more clean.

Edit:
`num.envs` is already an int ;)

Reply:
ok, but I would assume mask should be set to True at some point, no ?

Reply:
`mask` is used to reset a specific env `state` but we are only using the first env. because of this we cane just reset all of the `state` with `state = None`

Reply:
I know it works in that very special case, so maybe at least put a comment.

Reply:
I'd be very confused if I saw code using `states` but not `mask`. The code already has a `dones` flag, it doesn't seem hard to initialize that earlier in the code and then pass it into `mask`.

I'll also note that setting `state = None` *might* have different semantics to `mask = True` even in the single-environment case, for more exotic algorithms like $\text{RL}^2$ that actually maintain state across episodes.

Reply:
ok this all sounds like it make sense. I have added masks again.